File: ios/Teleport/TeleportAPI.swift
import Foundation

enum APIError: LocalizedError {
    case invalidURL
    case http(Int, String)
    case decode(Error)
    case network(Error)

    var errorDescription: String? {
        switch self {
        case .invalidURL: return "Неверный URL сервера"
        case .http(let code, let msg): return msg.isEmpty ? "Ошибка \(code)" : msg
        case .decode: return "Ошибка разбора ответа"
        case .network(let e): return e.localizedDescription
        }
    }
}

final class TeleportAPI {
    static let shared = TeleportAPI()
    private let decoder: JSONDecoder = {
        let d = JSONDecoder()
        return d
    }()
    private let encoder = JSONEncoder()

    private func url(_ path: String) throws -> URL {
        guard let u = URL(string: AppConfig.apiBaseURL + path.trimmingCharacters(in: CharacterSet(charactersIn: "/"))) else {
            throw APIError.invalidURL
        }
        return u
    }

    private func request<T: Decodable>(
        _ path: String,
        method: String = "GET",
        token: String? = nil,
        body: Encodable? = nil
    ) async throws -> T {
        var req = URLRequest(url: try url(path))
        req.httpMethod = method
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if let token { req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization") }
        if let body {
            req.httpBody = try encoder.encode(AnyEncodable(body))
        }
        do {
            let (data, res) = try await URLSession.shared.data(for: req)
            guard let http = res as? HTTPURLResponse else { throw APIError.http(0, "Нет ответа") }
            if http.statusCode >= 400 {
                let msg = (try? JSONDecoder().decode([String: String].self, from: data))?["detail"]
                    ?? String(data: data, encoding: .utf8)
                    ?? ""
                throw APIError.http(http.statusCode, msg)
            }
            if T.self == EmptyResponse.self, data.isEmpty {
                return EmptyResponse() as! T
            }
            do {
                return try decoder.decode(T.self, from: data)
            } catch {
                throw APIError.decode(error)
            }
        } catch let e as APIError {
            throw e
        } catch {
            throw APIError.network(error)
        }
    }

    func login(username: String, password: String) async throws -> AuthResponse {
        try await request("auth/login/username", method: "POST", body: UsernameLoginRequest(username: username, password: password))
    }

    func register(displayName: String, username: String, password: String) async throws -> AuthResponse {
        try await request("auth/register/web", method: "POST", body: WebRegisterRequest(displayName: displayName, username: username, password: password))
    }

    func sendSms(phone: String) async throws -> SmsSendResponse {
        try await request("auth/sms/send", method: "POST", body: SmsSendRequest(phone: phone))
    }

    func verifySms(phone: String, code: String) async throws -> SmsVerifyResponse {
        try await request("auth/sms/verify", method: "POST", body: SmsVerifyRequest(phone: phone, code: code))
    }

    func me(token: String) async throws -> UserDto {
        try await request("users/me", token: token)
    }

    func updateMe(token: String, body: UpdateProfileRequest) async throws -> UserDto {
        try await request("users/me", method: "PATCH", token: token, body: body)
    }

    func searchUsers(token: String, q: String) async throws -> [UserDto] {
        try await request("users/search?q=\(q.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? q)", token: token)
    }

    func listChats(token: String) async throws -> [ChatListItem] {
        try await request("chats", token: token)
    }

    func openChat(token: String, otherUserId: String) async throws -> OpenChatResponse {
        try await request("chats/open", method: "POST", token: token, body: OpenChatRequest(otherUserId: otherUserId))
    }

    func messages(token: String, chatId: String) async throws -> [MessageDto] {
        let page: MessagesPage = try await request("chats/\(chatId)/messages?since=0&limit=300", token: token)
        return page.messages
    }

    func sendMessage(token: String, chatId: String, text: String) async throws -> MessageDto {
        try await request("messages/send", method: "POST", token: token, body: SendMessageRequest(chatId: chatId, type: "TEXT", text: text))
    }
}

struct EmptyResponse: Codable {}

private struct AnyEncodable: Encodable {
    private let encodeFunc: (Encoder) throws -> Void
    init<T: Encodable>(_ wrapped: T) {
        encodeFunc = wrapped.encode
    }
    func encode(to encoder: Encoder) throws { try encodeFunc(encoder) }
}

File: android/app/src/main/java/com/teleport/messenger/data/api/TeleportApi.kt
package com.teleport.messenger.data.api

import com.teleport.messenger.BuildConfig
import com.jakewharton.retrofit2.converter.kotlinx.serialization.asConverterFactory
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MultipartBody
import okhttp3.OkHttpClient
import okhttp3.RequestBody.Companion.asRequestBody
import retrofit2.Retrofit
import retrofit2.http.*
import retrofit2.http.Part
import java.io.File
import java.util.concurrent.TimeUnit

interface TeleportApi {
    @POST("auth/register")
    suspend fun register(@Body body: AuthRequest): AuthResponse

    @POST("auth/login")
    suspend fun login(@Body body: AuthRequest): AuthResponse

    @POST("auth/login/username")
    suspend fun loginByUsername(@Body body: UsernameLoginRequest): AuthResponse

    @POST("auth/register/web")
    suspend fun registerWeb(@Body body: WebRegisterRequest): AuthResponse

    @POST("auth/qr")
    suspend fun qrLogin(@Body body: QrAuthRequest): AuthResponse

    @POST("auth/recover")
    suspend fun recover(@Body body: AuthRequest): AuthResponse

    @POST("auth/sms/send")
    suspend fun sendSms(@Body body: SmsSendRequest): SmsSendResponse

    @POST("auth/sms/verify")
    suspend fun verifySms(@Body body: SmsVerifyRequest): SmsVerifyResponse

    @POST("devices/fcm")
    suspend fun registerFcm(@Header("Authorization") token: String, @Body body: FcmTokenRequest)

    @Multipart
    @POST("media/upload")
    suspend fun uploadMedia(@Header("Authorization") token: String, @Part file: MultipartBody.Part): MediaUploadResponse

    @POST("messages/send")
    suspend fun sendMessage(@Header("Authorization") token: String, @Body body: SendMessageRequest): MessageDto

    @PATCH("messages/{id}")
    suspend fun editMessage(
        @Header("Authorization") token: String,
        @Path("id") id: String,
        @Body body: EditMessageRequest,
    ): MessageDto

    @POST("messages/{id}/forward")
    suspend fun forwardMessage(
        @Header("Authorization") token: String,
        @Path("id") id: String,
        @Body body: ForwardMessageRequest,
    ): MessageDto

    @GET("messages/sync")
    suspend fun syncMessages(@Header("Authorization") token: String, @Query("since") since: Long): SyncResponse

    @GET("users/search")
    suspend fun searchUsers(@Header("Authorization") token: String, @Query("q") query: String): List<UserDto>

    @GET("users/me")
    suspend fun getMe(@Header("Authorization") token: String): UserDto

    @PATCH("users/me")
    suspend fun updateMe(@Header("Authorization") token: String, @Body body: UpdateProfileRequest): UserDto

    @GET("users/{id}")
    suspend fun getUser(@Header("Authorization") token: String, @Path("id") id: String): UserDto

    @POST("chats/open")
    suspend fun openChat(@Header("Authorization") token: String, @Body body: OpenChatRequest): OpenChatResponse

    @GET("chats")
    suspend fun listChats(@Header("Authorization") token: String): List<ChatListItemDto>

    @GET("users/username/{username}/available")
    suspend fun checkUsername(@Path("username") username: String, @Query("exclude") exclude: String? = null): UsernameCheckResponse

    @GET("admin/check")
    suspend fun adminCheck(@Header("Authorization") token: String): AdminCheckDto

    @GET("admin/stats")
    suspend fun adminStats(@Header("Authorization") token: String): AdminStatsDto
}

object ApiClient {
    val json = Json { ignoreUnknownKeys = true; isLenient = true }

    val httpClient: OkHttpClient = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .writeTimeout(60, TimeUnit.SECONDS)
        .build()

    val api: TeleportApi = Retrofit.Builder()
        .baseUrl(BuildConfig.API_BASE_URL)
        .client(httpClient)
        .addConverterFactory(json.asConverterFactory("application/json".toMediaType()))
        .build()
        .create(TeleportApi::class.java)

    fun wsUrl(token: String): String {
        val base = BuildConfig.API_BASE_URL.removeSuffix("/")
        val wsBase = base.replace("http://", "ws://").replace("https://", "wss://")
        return "$wsBase/ws?token=$token"
    }

    suspend fun uploadFile(token: String, file: File, mime: String): String {
        val part = MultipartBody.Part.createFormData(
            "file",
            file.name,
            file.asRequestBody(mime.toMediaType()),
        )
        val path = api.uploadMedia("Bearer $token", part).url
        val base = BuildConfig.API_BASE_URL.removeSuffix("/")
        return if (path.startsWith("http")) path else "$base$path"
    }
}
